server/services/v1/database/query_index.go

```diff
@@ -0,0 +1,294 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/buger/jsonparser"
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/keys"
+	"github.com/tigrisdata/tigris/schema"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/store/kv"
+	"github.com/tigrisdata/tigris/value"
+	"golang.org/x/text/collate"
+	"golang.org/x/text/language"
+)
+
+const (
+	MAX_STRING_LEN = 64
+)
+
+type IndexIterator struct {
+	tx   transaction.Tx
+	it   kv.Iterator
+	ctx  context.Context
+	keys []keys.Key
+	err  error
+}
+
+type IndexRow struct {
+	docId []byte
+}
+
+func (iter *IndexIterator) Next(row *kv.KeyValue) bool {
+	if iter.err != nil {
+		return false
+	}
+
+	if iter.it.Next(row) {
+		return true
+	}
+
+	return false
+}
+
+func (it *IndexIterator) Interrupted() error { return it.err }
+
+type FieldIndex struct {
+	value value.Value
+	name  string
+	pos   int
+}
+
+func newFieldIndex(dataType schema.FieldType, name string, rawValue []byte, pos int) (*FieldIndex, error) {
+	if dataType == schema.StringType {
+		rawValue = collateString(rawValue)
+	}
+
+	value, err := value.NewValue(dataType, rawValue)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &FieldIndex{
+		value,
+		name,
+		pos,
+	}, nil
+}
+
+type QueryIndexInfo struct {
+	Rows int64
+	Size int64
+}
+
+func genRowCountKey(coll *schema.DefaultCollection) keys.Key {
+	return keys.NewKey(coll.EncodedName, coll.Indexes.QueryIndex.Name, "_info", "count")
+}
+
+func genIndexSizeKey(coll *schema.DefaultCollection) keys.Key {
+	return keys.NewKey(coll.EncodedName, coll.Indexes.QueryIndex.Name, "_info", "size")
+}
+
+type QueryIndex struct {
+	objectDepth int
+}
+
+func NewQueryIndexStore() *QueryIndex {
+
+	return &QueryIndex{
+		objectDepth: 4,
+	}
+}
+
+func (q *QueryIndex) ScanIndex(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, start keys.Key, end keys.Key) (*IndexIterator, error) {
+
+	it, err := tx.ReadRange(ctx, start, end, false)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &IndexIterator{
+		tx:   tx,
+		it:   it,
+		keys: []keys.Key{start, end},
+	}, nil
+}
+
+func (q *QueryIndex) IndexInfo(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection) (*QueryIndexInfo, error) {
+	rowCountKey := genRowCountKey(coll)
+	indexSizeKey := genIndexSizeKey(coll)
+
+	rows, err := tx.AtomicRead(ctx, rowCountKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	size, err := tx.AtomicRead(ctx, indexSizeKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &QueryIndexInfo{
+		rows,
+		size,
+	}, nil
+}
+
+func (q *QueryIndex) Index(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, doc []byte, id []interface{}) error {
```

Comment:
Should we refer to `id` as `primaryKey`, probably be more readable in this context that indexes are pointing to the primary key?

Reply:
Good point. I've fixed it.